web3.js/src/layout.ts

```diff
@@ -32,44 +55,54 @@ export const rustString = (property: string = 'string') => {
   const _decode = rsl.decode.bind(rsl);
   const _encode = rsl.encode.bind(rsl);
 
-  rsl.decode = (buffer: any, offset: any) => {
-    const data = _decode(buffer, offset);
-    return data['chars'].toString('utf8');
+  const rslShim = rsl as unknown as IRustStringShim;
+
+  rslShim.decode = (b: Uint8Array, offset?: number) => {
+    const data = _decode(b, offset);
+    return data['chars'].toString();
   };
 
-  rsl.encode = (str: any, buffer: any, offset: any) => {
+  rslShim.encode = (str: string, b: Uint8Array, offset?: number) => {
     const data = {
       chars: Buffer.from(str, 'utf8'),
     };
-    return _encode(data, buffer, offset);
+    return _encode(data, b, offset);
   };
 
-  (rsl as any).alloc = (str: any) => {
+  rslShim.alloc = (str: string) => {
     return (
       BufferLayout.u32().span +
       BufferLayout.u32().span +
       Buffer.from(str, 'utf8').length
     );
   };
 
-  return rsl;
+  return rslShim;
 };
 
 /**
  * Layout for an Authorized object
  */
 export const authorized = (property: string = 'authorized') => {
-  return BufferLayout.struct(
-    [publicKey('staker'), publicKey('withdrawer')],
-    property,
-  );
+  return BufferLayout.struct<
+    Readonly<{
+      staker: Uint8Array;
+      withdrawer: Uint8Array;
+    }>
+  >([publicKey('staker'), publicKey('withdrawer')], property);
```

Comment:
It definitely weirds me out that `publicKey` is decoding to just 32 bytes instead of a `PublicKey` instance. I guess maybe that's a reason _not_ to use buffer-layout-utils here, since it would be a breaking change.

Reply:
I haven't looked deeply into it, but I _presume_ that the way to achieve this is to create custom subclasses of `Layout` that serialize and deserialize to `PublicKey`, but read and write `Uint8Array` to the actual underlying buffer.

Curious, do you think this could be reimplemented using a subclass of `Layout` rather than patching `encode/decode` at runtime?

https://github.com/solana-labs/buffer-layout-utils/blob/master/src/web3.ts

Reply:
> I haven't looked deeply into it, but I _presume_ that the way to achieve this is to create custom subclasses of `Layout` that serialize and deserialize to `PublicKey`, but read and write `Uint8Array` to the actual underlying buffer.
> 
> Curious, do you think this could be reimplemented using a subclass of `Layout` rather than patching `encode/decode` at runtime?

Hmm possibly, it is a bit hacky.